File: static/js/restricted/titlepage/titlepage.js
/**
 * Rawscripts - Screenwriting Software
 * Copyright (C) Ritchie Wilson
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


window['closeTitlePage']=closeTitlePage;
window['save']=save;
window['update']=update;


//Set up global events.
//First Clicking on checkbox
var arr = goog.dom.getElementsByClass('checkbox')
for (i in arr){
	if(arr[i].type=='checkbox')goog.events.listen(arr[i], goog.events.EventType.CLICK, update)
}
// Listen for keyup events in text boxes
var arr = goog.dom.getElementsByClass('textbox')
for (i in arr){
	if(arr[i].nodeName=='INPUT' || arr[i].nodeName=='TEXTAREA')goog.events.listen(arr[i], goog.events.EventType.KEYUP, update)
}
// Listen for resize events because I suck at css. Suck suck suck.
var vsm = new goog.dom.ViewportSizeMonitor();
goog.events.listen(vsm, goog.events.EventType.RESIZE, setElementSizes);
// then initial resize
setElementSizes();

function setElementSizes(){
    var s = goog.dom.getViewportSize();
    var c = goog.style.getSize(goog.dom.getElement('controls'));
    goog.dom.getElement('pageContainer').style.width = (s.width - c.width ) + "px";
    goog.dom.getElement('page').style.visibility = 'visible';
}

function closeTitlePage(){
    window.close()
}
function save(){
    var resource_id=window.location.href.split('=')[1];
    if (resource_id=='Demo'){
        alert("Sorry, but you'll have to login to use these functions!");
        return;
    }
    var inputs = [
        ['title', 'title_input'],
        ['authorOne', 'authorOne_input'],
        ['authorTwo', 'authorTwo_input'],
        ['authorTwoChecked', 'authorTwo_checkbox'],
        ['authorThree', 'authorThree_input'],
        ['authorThreeChecked', 'authorThree_checkbox'],
        ['based_on', 'based_on_input'],
        ['based_onChecked', 'based_on_checkbox'],
        ['address', 'address_input'],
        ['addressChecked', 'address_checkbox'],
        ['phone', 'phone_input'],
        ['phoneChecked', 'phone_checkbox'],
        ['cell', 'cell_input'],
        ['cellChecked', 'cell_checkbox'],
        ['email', 'email_input'],
        ['emailChecked', 'email_checkbox'],
        ['registered', 'registered_input'],
        ['registeredChecked', 'registered_checkbox'],
        ['other', 'other_input'],
        ['otherChecked', 'other_checkbox']
    ];
    var postData = '';
    for (i in inputs){
        var elem = document.getElementById(inputs[i][1]);
        var value = '';
        if (elem.type == 'checkbox')
            value = (elem.checked ? 'checked' : '');
        else if (elem.value != undefined){
            value = elem.value;
            var re = new RegExp( "\\n", "g" );
            value = value.replace(re,'LINEBREAK');
        }
        if (postData != '')
            postData += '&';
        postData += inputs[i][0] + '=' + encodeURIComponent(value);
    }

	postData+='&resource_id='+resource_id;
	goog.net.XhrIo.send('/titlepagesave',
		function(e){
			goog.dom.getElement('saveButton').value='Saved'
		},
		'POST',
		postData
	);
    goog.dom.getElement('saveButton').disabled=true;
    goog.dom.getElement('saveButton').value="Saving..."
}


function update(){
    goog.dom.getElement('saveButton').value="Save";
    goog.dom.getElement('saveButton').disabled=false;
    goog.dom.getElement('title').innerHTML="";
    goog.dom.getElement('title').appendChild(document.createTextNode(goog.dom.getElement("title_input").value));
    goog.dom.getElement('authorOne').innerHTML="";
    goog.dom.getElement('authorOne').appendChild(document.createTextNode(goog.dom.getElement("authorOne_input").value));
    var c=document.getElementsByTagName('input')
    for(i in c){
        if (c[i].type=='checkbox'){
            var check = (c[i].checked==true ? true : false);
            var n = c[i].parentNode;
            n=n.previousSibling;
            if(n.nodeName=='#text')n=n.previousSibling;
            n=n.firstChild;
            if(n.nodeName=='#text')n=n.nextSibling;
            if(n.value=="" && !check)n.value="none";
            var data = n.value;
            if (c[i].id=="address_checkbox" || c[i].id=="based_on_checkbox"){
                var re = new RegExp( "\\n", "g" );
                data=data.replace(re,'<br>');
                goog.dom.getElement(c[i].id.replace('_checkbox',"")).innerHTML=data;
            }
            else{
                goog.dom.getElement(c[i].id.replace('_checkbox',"")).innerHTML="";
                goog.dom.getElement(c[i].id.replace('_checkbox',"")).appendChild(document.createTextNode(data));
            }
            if(check){
                goog.dom.getElement(c[i].id.replace('_checkbox',"_input")).disabled=false;
                goog.dom.getElement(c[i].id.replace('_checkbox',"")).style.visibility='visible';
            }
            else{
                goog.dom.getElement(c[i].id.replace('_checkbox',"_input")).disabled=true;
                goog.dom.getElement(c[i].id.replace('_checkbox',"")).style.visibility='hidden';
            }
        }
    }
}

File: appengine/titlepage.py
# Rawscripts - Screenwriting Software
# Copyright (C) Ritchie Wilson
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


import StringIO, os, cgi, re
import wsgiref.handlers
from google.appengine.api import users
from google.appengine.ext import webapp
from google.appengine.ext import db
from google.appengine.ext.webapp import template
from google.appengine.ext.webapp.util import run_wsgi_app
import config
from models import TitlePageData

from utils import ownerPermission, get_template_path

class TitlePage(webapp.RequestHandler):
    def get(self):
        resource_id = self.request.get('resource_id')
        if resource_id == "Demo":
            p = "Duck Soup"
        else:
            user = users.get_current_user()
            if not user:
                self.redirect('/')
                return
            p = ownerPermission(resource_id)
        if p == False:
            return

        if resource_id == "Demo":
            template_values = { 'sign_out': "/" }
            template_values['user'] = "test@example.com"

        else:
            template_values = { 'sign_out': users.create_logout_url('/') }
            template_values['user'] = users.get_current_user().email()

        if resource_id == "Demo":
            template_values = {
                'title' : "Duck Soup",
                'authorOne' : "Arthur Sheekman",
                'authorTwo' : "Harry Ruby",
                'authorTwoChecked' : "checked",
                'authorThree' : "Bert Kalmar",
                'authorThreeChecked': "checked",
                'based_on' : "none",
                'based_onChecked' : "",
                'address' : "183 E. 93rd St\nSuite 9\nNY, NY",
                'addressChecked' : "checked",
                'phone' : "212-555-5555",
                'phoneChecked' : "checked",
                'cell' : "",
                'cellChecked' : "",
                'email' : "test@example.com",
                'emailChecked' : "checked",
                'registered': "",
                'registeredChecked' : "",
                'other' : "",
                'otherChecked' : ""
            }

        else:
            r = TitlePageData.get_or_create(resource_id, p)
            template_values = {
                'title' : r.title,
                'authorOne' : r.authorOne,
                'authorTwo' : r.authorTwo,
                'authorTwoChecked' : r.authorTwoChecked,
                'authorThree' : r.authorThree,
                'authorThreeChecked': r.authorThreeChecked,
                'based_on' : r.based_on.replace("LINEBREAK", '\n'),
                'based_onChecked' : r.based_onChecked,
                'address' : r.address.replace("LINEBREAK", '\n'),
                'addressChecked' : r.addressChecked,
                'phone' : r.phone,
                'phoneChecked' : r.phoneChecked,
                'cell' : r.cell,
                'cellChecked' : r.cellChecked,
                'email' : r.email,
                'emailChecked' :r.emailChecked,
                'registered': r.registered,
                'registeredChecked' : r.registeredChecked,
                'other' : r.other,
                'otherChecked' : r.otherChecked
            }

        template_values['MODE'] = config.MODE
        template_values['GA'] = config.GA
        path = get_template_path('html/titlepage.html')
        self.response.headers['Content-Type'] = 'text/html'
        self.response.out.write(template.render(path, template_values))

class SaveTitlePage (webapp.RequestHandler):
    def post(self):
        resource_id = self.request.get('resource_id')
        if resource_id == "Demo":
            return
        title = ownerPermission(resource_id)
        if title == False:
            return

        i = TitlePageData().get_or_create(resource_id, title)

        i.resource_id = resource_id
        i.title = self.request.get('title')
        i.authorOne = self.request.get('authorOne')
        i.authorTwo = self.request.get('authorTwo')
        i.authorTwoChecked = self.request.get('authorTwoChecked')
        i.authorThree = self.request.get('authorThree')
        i.authorThreeChecked = self.request.get('authorThreeChecked')
        i.based_on = self.request.get('based_on')
        i.based_onChecked = self.request.get('based_onChecked')
        i.address = self.request.get('address')
        i.addressChecked = self.request.get('addressChecked')
        i.phone = self.request.get('phone')
        i.phoneChecked = self.request.get('phoneChecked')
        i.cell = self.request.get('cell')
        i.cellChecked = self.request.get('cellChecked')
        i.email = self.request.get('email')
        i.emailChecked = self.request.get('emailChecked')
        i.registered = self.request.get('registered')
        i.registeredChecked = self.request.get('registeredChecked')
        i.other = self.request.get('other')
        i.otherChecked = self.request.get('otherChecked')
        i.put()

        self.response.headers['Content-Type']='text/plain'
        self.response.out.write('1')

def main():
    application = webapp.WSGIApplication([('/titlepage', TitlePage),
                                          ('/titlepagesave', SaveTitlePage)],
                                         debug=True)

    run_wsgi_app(application)

if __name__ == '__main__':
    main()
